dpctl/tensor/_testing.py

```diff
@@ -0,0 +1,144 @@
+#                      Data Parallel Control (dpctl)
+#
+# Copyright 2020-2023 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+import dpctl.tensor as dpt
+import dpctl.utils as du
+
+from ._manipulation_functions import _broadcast_shape_impl
+from ._type_utils import _to_device_supported_dtype
+
+
+def _allclose_complex_fp(z1, z2, atol, rtol, equal_nan):
+    z1r = dpt.real(z1)
+    z1i = dpt.imag(z1)
+    z2r = dpt.real(z2)
+    z2i = dpt.imag(z2)
+    if equal_nan:
+        check1 = dpt.all(dpt.isnan(z1r) == dpt.isnan(z2r)) and dpt.all(
+            dpt.isnan(z1i) == dpt.isnan(z2i)
+        )
+    else:
+        check1 = (
+            dpt.logical_not(dpt.any(dpt.isnan(z1r)))
+            and dpt.logical_not(dpt.any(dpt.isnan(z1i)))
+        ) and (
+            dpt.logical_not(dpt.any(dpt.isnan(z2r)))
+            and dpt.logical_not(dpt.any(dpt.isnan(z2i)))
+        )
+    if not check1:
+        return check1
+    mr = dpt.isinf(z1r)
+    mi = dpt.isinf(z1i)
+    check2 = dpt.all(mr == dpt.isinf(z2r)) and dpt.all(mi == dpt.isinf(z2i))
+    if not check2:
+        return check2
+    check3 = dpt.all(z1r[mr] == z2r[mr]) and dpt.all(z1i[mi] == z2i[mi])
+    if not check3:
+        return check3
+    mr = dpt.isfinite(z1r)
+    mi = dpt.isfinite(z1i)
+    mv1 = z1r[mr]
+    mv2 = z2r[mr]
+    check4 = dpt.all(
+        dpt.abs(mv1 - mv2)
+        < atol + rtol * dpt.maximum(dpt.abs(mv1), dpt.abs(mv2))
+    )
+    if not check4:
+        return check4
+    mv1 = z1i[mi]
+    mv2 = z2i[mi]
+    check5 = dpt.all(
+        dpt.abs(mv1 - mv2)
+        < atol + rtol * dpt.maximum(dpt.abs(mv1), dpt.abs(mv2))
+    )
+    return check5
+
+
+def _allclose_real_fp(r1, r2, atol, rtol, equal_nan):
+    if equal_nan:
+        check1 = dpt.all(dpt.isnan(r1) == dpt.isnan(r2))
+    else:
+        check1 = dpt.logical_not(dpt.any(dpt.isnan(r1))) and dpt.logical_not(
+            dpt.any(dpt.isnan(r2))
+        )
+    if not check1:
+        return check1
+    mr = dpt.isinf(r1)
+    check2 = dpt.all(mr == dpt.isinf(r2))
+    if not check2:
+        return check2
+    check3 = dpt.all(r1[mr] == r2[mr])
+    if not check3:
+        return check3
+    m = dpt.isfinite(r1)
+    mv1 = r1[m]
+    mv2 = r2[m]
+    check4 = dpt.all(
+        dpt.abs(mv1 - mv2)
+        < atol + rtol * dpt.maximum(dpt.abs(mv1), dpt.abs(mv2))
+    )
```

Comment:
According to Numpy, these conditions should be ``<=``

See here
```
In [1]: import dpctl.tensor as dpt, numpy as np

In [2]: atol = 1e-5

In [3]: x = dpt.zeros(10, dtype="f4")

In [4]: y = dpt.asarray(atol, dtype="f4")

In [5]: dpt.allclose(x, y, atol=atol, rtol=0)
Out[5]: usm_ndarray(False)

In [6]: x_np, y_np = dpt.asnumpy(x), dpt.asnumpy(y)

In [7]: np.allclose(x_np, y_np, atol=atol, rtol=0)
Out[7]: True

In [8]: dpt.abs(x - y) < (atol + 0 * dpt.maximum(dpt.abs(x), dpt.abs(y)))
Out[8]:
usm_ndarray([False, False, False, False, False, False, False, False,
             False, False])

In [9]: dpt.abs(x - y) <= (atol + 0 * dpt.maximum(dpt.abs(x), dpt.abs(y)))
Out[9]:
usm_ndarray([ True,  True,  True,  True,  True,  True,  True,  True,
              True,  True])
```

Reply:
Also, I think it would be a good idea to add a test for the behavior of the ``atol`` and ``rtol`` parameters, much like the above.

Reply:
Test added, and `<` replaced with `<=`.